recycle.1
.TH RECYCLE 1 "March 2025" "Version 1.0" "Recycle Script Manual"

.SH NAME
recycle \- move files to a recycle bin directory with Git tracking

.SH SYNOPSIS
.B recycle
[\fIoptions\fR] <file1> <file2> ...

.SH DESCRIPTION
The \fBrecycle\fR script moves specified files into a designated recycle bin directory while tracking changes using Git. It is designed for preserving temporary or one-time-use files that may be useful later.

.SH OPTIONS

.TP
.B \-b, \-\-branch <branchname>
Switch to the specified branch named \fIbranchname\fR in the recycle bin repository before recycling any files. If the branch does not exist, it will be created based on the \fImaster\fR branch. The specified branch will remain active after the operation. If this option is not used, the currently active branch will be used.

.TP
.B \-cp, \-\-cherry-pick, \-\-copy <hash_of_commit>
Copy the recycled files from a specified commit into the active branch. This option allows you to cherry-pick commits from another branch into your current branch. It is particularly useful for maintaining a clean installation by copying commits from other branches (e.g., from an upstream repository) and then removing the original branch if desired.

.TP
.B \-f, \-\-force
Enable force mode for subsequent operations. When used with --push, it forces the push operation even if it results in a non-fast-forward update. For file operations, it allows overwriting of existing files in the recycle bin. Use with extreme caution as it can overwrite remote changes and local recycled files, potentially causing data loss. NOTE: When using --force with --push, the order is significant. To perform a force push, use --force before --push.

.TP
.B \-ls, \-\-list
List files in the recycle bin.
.TP
.B \-ll, \-\-llist
List files in the recycle bin in the long listing format.

.TP
.B \-\-log[=<filename>]
Show the git commit log for the recycle bin, or a specific file within if specified.

.TP
\fB-m, --message\fR "commit message"
Specify a custom commit message for Git. If not provided, a default message will be generated using the current date and time. The list of files is always included in the commit message body.

.TP
.B \-\-push[=<remote>]
Uploads the current branch of the recycle bin repository to the specified remote. If no remote is specified, it defaults to 'origin'. This option allows you to share your recycled changes with a remote repository, making them accessible to other users or systems. It's useful for backing up your recycled files or synchronizing changes across different environments.

Note: Ensure you have the necessary permissions to push to the specified remote.

.SH USAGE EXAMPLES
.TP
Recycle a single file:
.B recycle /path/to/file.txt

.TP
Recycle multiple files with a custom commit message:
.B recycle -m \(dqscraped example.com\(dq /path/to/scrape_example.sh /path/to/post_process_example.sh


.TP
Recycle multiple files with the default commit message:
.B recycle /path/to/file1.txt /path/to/file2.txt

.TP
Initialize a new personal branch:
.B recycle -b=$USER

.TP
Recycle files into a private branch:
.B recycle -b=private /path/to/sensitive/file1.txt /path/to/unshared/file2.txt

.TP
Switch to a personal branch and copy the specified commit into it. This allows you to integrate specific recycled files without merging entire branches.
.B recycle -b=$USER -cp=<hash_of_commit>

.TP
Synchronize the remote repository, by pushing a personal branch to origin and a private branch to a private remote. Note: the recycle bin will be in the private branch afterward.
.B recycle -b=$USER --push -b=private --push=<non_public_remote>

.SH DEFAULT COMMIT MESSAGE FORMAT
.TP
The head line defaults to a timestamp in the format:
.B [YYYY-MM-DD-HHmm.ss]

.TP
In all cases, the content paragraph will contain the list of files:
.B /original/file/path/of/first/input.file
.br
.B ...
.br
.B /original/file/path/of/last/input.file

.TP
File paths are resolved to their absolute form using realpath, handling relative paths and symbolic links. However, in the commit message, home directory paths are anonymized by removing the \fI/home/username/\fR prefix.

.SH REPOSITORY STRUCTURE
The main branch contains development for the \fBrecycle\fR script itself. Other branches store recycled items for individual users, systems, or purposes.

.SH LICENSES AND SHARING PHILOSOPHY
The \fBrecycle\fR script is licensed under the Artistic License 2.0.
Recycled items are considered public domain unless explicitly stated otherwise.

.SH AUTHOR
Hildigerr Vergaray
.br
https://github.com/hildigerr/recycle.bin


recycle.sh
#!/bin/bash

# Default recycle bin directory
RECYCLE_BIN_DIR="${RECYCLE_BIN_DIR:-$HOME/.local/share/recycle.bin}"

# Default commit message head line
COMMIT_MESSAGE="[$(date +"%Y-%m-%d-%H%M.%S")]"

# Silence pushd and popd
pushd () {
  command pushd "$@" > /dev/null
}
popd () {
  command popd "$@" > /dev/null
}

# Check if RECYCLE_BIN_DIR exists, if not create it and initialize git
if [ ! -d "$RECYCLE_BIN_DIR" ]; then
  mkdir -p "$RECYCLE_BIN_DIR"
  pushd "$RECYCLE_BIN_DIR" || exit 1
  git init
  popd
fi

branch() {
  BRANCH=$1
  pushd "$RECYCLE_BIN_DIR"
  git checkout "$BRANCH" 2>/dev/null || \
  git checkout -b "$BRANCH" master || exit 1
  popd 
}

cherry_pick() {
  HASH=$1
  pushd "$RECYCLE_BIN_DIR"
  git cherry-pick $HASH || exit 1
  popd 
}

git_log() {
  pushd "$RECYCLE_BIN_DIR" || exit 1
  git log --all --oneline --decorate --graph $1
  popd
}

git_push() {
  REMOTE=$1
  pushd "$RECYCLE_BIN_DIR"
  git push $FORCE_P "$REMOTE" HEAD || exit 1
  popd 
}

# Parse arguments
FILES=()
while [[ "$#" -gt 0 ]]; do
  case $1 in
    --branch|-b) branch "$2"; shift ;;
    --branch=*|-b=*) branch "${1#*=}" ;;
    --cherry-pick|--copy|-cp) cherry_pick "$2"; shift ;;
    --cherry-pick=*|--copy=*|-cp=*) cherry_pick "${1#*=}" ;;
    --force|-f) FORCE_P="--force" ;;
    --list|-ls) pushd "$RECYCLE_BIN_DIR" && ls && popd ;;
    --llist|-ll) pushd "$RECYCLE_BIN_DIR" && ls -l && popd ;;
    --log) git_log ;;
    --log=*) git_log "${1#*=}" ;;
    --message|-m) COMMIT_MESSAGE="$2"; shift ;;
    --message=*|-m=*) COMMIT_MESSAGE="${1#*=}" ;;
    --push) git_push origin ;;
    --push=*) git_push "${1#*=}" ;;
    *) if [ -e "$1" ]; then FILES+=("$(realpath -m -- "$1")")
       else echo "Warning: File '$1' does not exist. Skipping."
       fi ;;
  esac
  shift
done

# Check if valid files are provided
if [ ${#FILES[@]} -eq 0 ]; then
  echo "Warning: Nothing to recycle. Done."
  exit
elif [ -z "$FORCE_P" ] ; then
  for FILE in "${FILES[@]}"; do
    BASENAME=$(basename "$FILE")
    if [ -e "$RECYCLE_BIN_DIR/$BASENAME" ]; then
      echo "Error: '$BASENAME' already exists in the recycle bin."
      echo "Operation aborted. Use --force to overwrite existing files."
      exit 1
    fi
  done
fi

# Move files to the recycle bin directory
for FILE in "${FILES[@]}"; do
  mv "$FILE" "$RECYCLE_BIN_DIR"
done

# Stage and commit changes with git
cd "$RECYCLE_BIN_DIR" || exit 1
git add .
git commit -m "$COMMIT_MESSAGE" -m "$(printf "%s\n" "${FILES[@]#/home/*/}")"

